homeassistant/components/ezviz/config_flow.py

```diff
@@ -0,0 +1,261 @@
+"""Config flow for ezviz."""
+import logging
+
+from pyezviz import EzvizClient
+from pyezviz.client import PyEzvizError
+from requests.exceptions import HTTPError
+import voluptuous as vol
+
+from homeassistant import exceptions
+from homeassistant.config_entries import CONN_CLASS_CLOUD_POLL, ConfigFlow, OptionsFlow
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_REGION,
+    CONF_TIMEOUT,
+    CONF_TYPE,
+    CONF_USERNAME,
+)
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import AbortFlow
+
+from .const import (  # pylint: disable=unused-import
+    ATTR_SERIAL,
+    ATTR_TYPE_CAMERA,
+    ATTR_TYPE_CLOUD,
+    CONF_FFMPEG_ARGUMENTS,
+    DEFAULT_CAMERA_USERNAME,
+    DEFAULT_FFMPEG_ARGUMENTS,
+    DEFAULT_REGION,
+    DEFAULT_TIMEOUT,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class EzvizConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Ezviz."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_CLOUD_POLL
+
+    async def _validate_and_create_auth(self, data):
+        """Validate the user input allows us to connect.
+
+        Data has the keys from data_schema with values provided by the user.
+        """
+        await self.async_set_unique_id(data[CONF_USERNAME])
+        self._abort_if_unique_id_configured()
+
+        # constructor does login call
+        client = EzvizClient(
+            data[CONF_USERNAME],
+            data[CONF_PASSWORD],
+            data.get(CONF_REGION, DEFAULT_REGION),
+            data.get(CONF_TIMEOUT, DEFAULT_TIMEOUT),
+        )
+
+        # Validate data by sending a login request to ezviz
+
+        try:
+            await self.hass.async_add_executor_job(client.login)
+
+        except (PyEzvizError, HTTPError) as err:
```

Comment:
Why is the library not catching the HTTPError and raising a library specific exception? That seems to be a recent change in the library. Preferably the library should only raise library specific exceptions.

Reply:
The library (pyezviz) does. Just thought I'll also cater for general connection errors like packet loss when the users tries to add the main account.  A HTTError would then just return unable to authenticate and prompt for credentials and region again.

The config flow also caters for cloud accounts in other regions. If someone enters the incorrect region code it could cause a HTTPError as the host portion of the url is derived from the region code. ("eu" example: apiieu.ezvizlife.com)

Reply:
It's ok to catch more exceptions but that should be done by the library and then library specific exceptions raised.

Reply:
Done. Using library specific exceptions for general connectivity exceptions and PyEzvizError for authentication failures and invalid data returned from api.